glusterd2/commands/volumes/volume-status.go

```diff
@@ -111,31 +114,45 @@ func volumeStatusHandler(w http.ResponseWriter, r *http.Request) {
 
 	txn := transaction.NewTxn(ctx)
 	defer txn.Cleanup()
-	volNodes := vol.Nodes()
 	txn.Steps = []*transaction.Step{
 		{
 			DoFunc: "vol-status.Check",
-			Nodes:  volNodes,
+			Nodes:  vol.Nodes(),
 		},
 	}
-
 	txn.Ctx.Set("volname", volname)
 
 	// Some nodes may not be up, which is okay.
 	txn.DontCheckAlive = true
 	txn.DisableRollback = true
 
+	// txn.Do() call is synchronous; try getting volume size info
+	// concurrently while the transaction runs.
+	ch := make(chan *api.SizeInfo)
+	go func() {
+		s, err := volumeUsage(vol.Name)
+		if err != nil {
+			logger.WithError(err).WithField("volume", volname).Error("Failed to get volume size info")
+		}
+		ch <- s
+	}()
+
 	err = txn.Do()
 	if err != nil {
-		logger.WithFields(log.Fields{
-			"error":  err.Error(),
-			"volume": volname,
-		}).Error("volumeStatusHandler: Failed to get volume status.")
+		logger.WithError(err).WithField("volume", volname).Error("Failed to get volume status")
 		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err.Error(), api.ErrCodeDefault)
 		return
 	}
 
-	result, err := createVolumeStatusResp(txn.Ctx, vol)
+	var size *api.SizeInfo
+	select {
+	case size = <-ch:
+		// should be ready by the time the txn is over
+	case <-time.After(time.Second * 2):
```

Comment:
I am not very sure about what exactly are we trying to achieve here. If we're defining a particular time out here, then with more number of bricks in a volume would a mount be able to succeed and provide back the statfs details?

Reply:
If glusterd2 isn't able to mount and get size information within certain period of time, we quit waiting. This is intended to handle scenarios where mount can be stuck or is taking too long to respond.

Reply:
There's still an open question. With a higher number of bricks, the mount might genuinely take some additional time. We may need to actually decide wisely on the timeout value we choose here (IMHO).